packages/webpack-cli/src/plugins/dotenv-webpack-plugin.ts

```diff
@@ -0,0 +1,253 @@
+    const { allowEmptyValues, safe } = this.#options;
+    const vars: EnvVariables = this.#initializeVars();
+
+    const { env, blueprint } = await this.#getEnvs(compilation);
+
+    Object.keys(blueprint).forEach((key) => {
+      const value = Object.prototype.hasOwnProperty.call(vars, key) ? vars[key] : env[key];
+
+      if (
+        (typeof value === "undefined" || value === null || (!allowEmptyValues && value === "")) &&
+        safe
+      ) {
+        compilation.errors.push(new Error(`Missing environment variable: ${key}`));
+      } else {
+        vars[key] = value;
+      }
+      if (safe) {
+        Object.keys(env).forEach((key) => {
+          if (!Object.prototype.hasOwnProperty.call(vars, key)) {
+            vars[key] = env[key];
+          }
+        });
+      }
+    });
+
+    return vars;
+  }
+
+  #initializeVars(): EnvVariables {
+    return this.#options.systemvars
+      ? Object.fromEntries(Object.entries(process.env).map(([key, value]) => [key, value ?? ""]))
+      : {};
+  }
+
+  merge(apply = {}, defaults = {}) {
+    Object.assign({}, defaults, apply);
+  }
+
+  parse(src: string, defaultSrc = "") {
+    const parsedSrc = dotenv.parse(src);
+    const parsedDefault = dotenv.parse(defaultSrc);
+    const parsed = {};
+    Object.assign(parsed, parsedDefault, parsedSrc);
+    return parsed;
+  }
+
+  async #getEnvs(compilation: any): Promise<{ env: EnvVariables; blueprint: EnvVariables }> {
+    const { paths, safe } = this.#options;
+
+    const env: EnvVariables = {};
+    let blueprint: EnvVariables = {};
+    if (paths) {
+      for (const path of paths) {
+        const fileContent = await this.#loadFile(path, compilation);
+        Object.assign(env, this.parse(fileContent));
+      }
+      blueprint = env;
+      if (safe) {
+        for (const path of paths) {
+          const exampleContent = await this.#loadFile(`${path}.example`, compilation);
+          Object.assign(env, this.parse(exampleContent));
+        }
+      }
+    } else {
+      compilation.errors.push(new Error(`No paths in config.`));
+    }
+    return { env, blueprint };
+  }
+
+  #formatData({
+    variables = {},
+    target,
+  }: {
+    variables: EnvVariables;
+    target: string;
+  }): Record<string, string> {
+    const { expand, prefixes } = this.#options;
+
+    const preprocessedVariables: EnvVariables = Object.keys(variables).reduce(
+      (obj: EnvVariables, key: string) => {
+        let value = variables[key];
+        if (expand) {
+          if (value.startsWith("\\$")) {
+            value = value.substring(1);
+          } else if (value.includes("\\$")) {
+            value = value.replace(/\\\$/g, "$");
+          } else {
+            value = interpolate(value, variables);
+          }
+        }
+        obj[key] = JSON.stringify(value);
+        return obj;
+      },
+      {},
+    );
+
+    const formatted: Record<string, string> = {};
+    prefixes?.forEach((prefix) => {
```

Comment:
Never do it, it is unreadable, please use `if (...) { /* code */ }`, same for all places